pkg/manifest/jsonmanifest/manifest.go

```diff
@@ -6,51 +6,94 @@ package jsonmanifest
 // It stores manifest entries in a map based on string keys.
-type JSONManifest struct {
-	Entries map[string]*JSONEntry `json:"entries,omitempty"`
+type jsonManifest struct {
+	entriesMu sync.RWMutex // mutex for accessing the entries map
+	entries   map[string]*jsonEntry
 }
 
-// NewManifest creates a new JSONManifest struct and returns a pointer to it.
-func NewManifest() *JSONManifest {
-	return &JSONManifest{
-		Entries: make(map[string]*JSONEntry),
+// NewManifest creates a new jsonManifest struct and returns a pointer to it.
+func NewManifest() manifest.Interface {
+	return &jsonManifest{
+		entries: make(map[string]*jsonEntry),
 	}
 }
 
 // Add adds a manifest entry to the specified path.
-func (m *JSONManifest) Add(path string, entry manifest.Entry) {
-	m.Entries[path] = NewEntry(entry.Reference(), entry.Name(), entry.Headers())
+func (m *jsonManifest) Add(path string, entry manifest.Entry) {
+	m.entriesMu.Lock()
+	defer m.entriesMu.Unlock()
+
+	m.entries[path] = &jsonEntry{
+		reference: entry.Reference(),
+		name:      entry.Name(),
+		headers:   entry.Headers(),
+	}
 }
 
 // Remove removes a manifest entry on the specified path.
-func (m *JSONManifest) Remove(path string) {
-	delete(m.Entries, path)
+func (m *jsonManifest) Remove(path string) {
+	m.entriesMu.Lock()
+	defer m.entriesMu.Unlock()
+
+	delete(m.entries, path)
 }
 
 // Entry returns a manifest entry if one is found in the specified path.
-func (m *JSONManifest) Entry(path string) (manifest.Entry, error) {
-	if entry, ok := m.Entries[path]; ok {
-		return entry, nil
+func (m *jsonManifest) Entry(path string) (manifest.Entry, error) {
+	m.entriesMu.RLock()
+	defer m.entriesMu.RUnlock()
+
+	entry, ok := m.entries[path]
+	if !ok {
+		return nil, manifest.ErrNotFound
 	}
 
-	return nil, manifest.ErrNotFound
+	// return a copy to prevent external modification
+	return NewEntry(entry.Reference(), entry.Name(), entry.Headers()), nil
+}
+
+// Length returns the amount of entries in the manifest.
+func (m *jsonManifest) Length() int {
+	m.entriesMu.RLock()
+	defer m.entriesMu.RUnlock()
+
+	return len(m.entries)
+}
+
+// exportManifest is a struct used for marshaling and unmarshaling jsonManifest structs.
+type exportManifest struct {
+	Entries map[string]*jsonEntry `json:"entries,omitempty"`
 }
 
-// MarshalBinary implements encoding.BinaryMarshaler.
-func (m *JSONManifest) MarshalBinary() (data []byte, err error) {
-	return json.Marshal(m)
+// MarshalBinary implements encoding.BinaryMarshaler
+func (m *jsonManifest) MarshalBinary() ([]byte, error) {
+	m.entriesMu.RLock()
+	defer m.entriesMu.RUnlock()
+
+	return json.Marshal(exportManifest{
+		Entries: m.entries,
+	})
 }
 
 // UnmarshalBinary implements encoding.BinaryUnmarshaler.
-func (m *JSONManifest) UnmarshalBinary(data []byte) error {
-	return json.Unmarshal(data, m)
+func (m *jsonManifest) UnmarshalBinary(b []byte) error {
+	m.entriesMu.Lock()
+	defer m.entriesMu.Unlock()
+
+	e := exportManifest{}
```

Comment:
do we need a `make(map[string]*jsonEntry)` statement here?

why/why not?

Reply:
i think we do, because of the json serializer attribute: ``json:"entries,omitempty"`` - it will omit the entry in case it is empty. i'm not sure about this 100% but again - classic for a unit test

Reply:
after some discussion with @pradovic, we ended up simplifying the marshaling of the `jsonEntry` struct.

since it is now private, we can have the fields as exported with the json serializer attribute set, which allows us to just go ahead with the `json.Marshal` and `json.Unmarshal` calls.